`oj t -M all`でエラーが出る

## Summary / 概要
`oj t -M all`でエラーが出る

## Steps to reproduce / 再現方法

1. `oj t -M all`

environments:
-   version: 11.4.0
- WSL2 上の Ubuntu 20.04

## Expected behavior / 期待される挙動
適切に表示される

## Actual behavior / 実際の挙動

``` console
[ERROR] 
Traceback (most recent call last):
  File "/home/xps15/.local/lib/python3.8/site-packages/onlinejudge_command/main.py", line 101, in main
    sys.exit(run_program(parsed, parser=parser))
  File "/home/xps15/.local/lib/python3.8/site-packages/onlinejudge_command/main.py", line 70, in run_program
    if not subcommand_test.run(args):
  File "/home/xps15/.local/lib/python3.8/site-packages/onlinejudge_command/subcommand/test.py", line 315, in run
    history += [test_single_case(name, paths['in'], paths.get('out'), args=args)]
  File "/home/xps15/.local/lib/python3.8/site-packages/onlinejudge_command/subcommand/test.py", line 253, in test_single_case
    status = display_result(proc, answer, memory, test_input_path, test_output_path, mle=args.mle, display_mode=DisplayMode(args.display_mode), does_print_input=args.print_input, silent=args.silent, match_result=match_result)
  File "/home/xps15/.local/lib/python3.8/site-packages/onlinejudge_command/subcommand/test.py", line 210, in display_result
    assert False
AssertionError
```


## Other notes / その他
- summary, diff は正しく表示されています
-   Will you try to create a pull request?
    -  no

--- a/onlinejudge_command/output_comparators.py
+++ b/onlinejudge_command/output_comparators.py
@@ -114,5 +114,5 @@ class SpecialJudge:
             logger.info('$ %s', command)
             info, proc = utils.exec_command(command)
         if not self.is_silent:
-            logger.info(utils.NO_HEADER + 'judge\'s output:\n%s', pretty_printers.make_pretty_large_file_content(info['answer'] or b'', limit=40, head=20, tail=10, bold=True))
+            logger.info(utils.NO_HEADER + 'judge\'s output:\n%s', pretty_printers.make_pretty_large_file_content(info['answer'] or b'', limit=40, head=20, tail=10))
         return proc.returncode == 0

--- a/onlinejudge_command/pretty_printers.py
+++ b/onlinejudge_command/pretty_printers.py
@@ -26,34 +26,71 @@ class _PrettyToken(NamedTuple):
     value: str
 
 
+def _tokenize_line(line: str) -> List[_PrettyToken]:
+    body = line.rstrip()
+    newline = line[len(body):]
+    tokens = []
+
+    # add the body of line
+    if body:
+        l = 0
+        while l < len(body):
+            r = l + 1
+            while r < len(body) and (body[l] in ' \t') == (body[r] in ' \t'):
+                r += 1
+            if body[l] in ' \t':
+                typ = _PrettyTokenType.WHITESPACE
+            else:
+                typ = _PrettyTokenType.BODY
+            tokens.append(_PrettyToken(typ, body[l:r]))
+            l = r
+
+    # add newlines
+    if newline:
+        if newline in ('\n', '\r\n'):
+            tokens.append(_PrettyToken(_PrettyTokenType.NEWLINE, newline))
+        else:
+            whitespace = newline.rstrip('\n')
+            newline = newline[len(whitespace):]
+            if whitespace:
+                tokens.append(_PrettyToken(_PrettyTokenType.WHITESPACE, whitespace))
+            tokens.append(_PrettyToken(_PrettyTokenType.HINT, '(trailing whitespace)'))
+            if newline:
+                tokens.append(_PrettyToken(_PrettyTokenType.NEWLINE, newline))
+
+    return tokens
+
+
+def _decode_with_recovery(content: bytes) -> Tuple[List[_PrettyToken], str]:
+    tokens = []
+    try:
+        text = content.decode()
+    except UnicodeDecodeError as e:
+        tokens.append(_PrettyToken(_PrettyTokenType.HINT, str(e)))
+        text = content.decode(errors='replace')
+    return tokens, text
+
+
+def _warn_if_empty(tokens: List[_PrettyToken]) -> List[_PrettyToken]:
+    if not tokens:
+        return [_PrettyToken(_PrettyTokenType.HINT, '(empty)')]
+    if tokens[-1][0] == _PrettyTokenType.BODY:
+        tokens.append(_PrettyToken(_PrettyTokenType.HINT, '(no trailing newline)'))
+    if all(token.type == _PrettyTokenType.NEWLINE for token in tokens):
+        tokens.append(_PrettyToken(_PrettyTokenType.HINT, '(only newline)'))
+    return tokens
+
+
 def _tokenize_large_file_content(*, content: bytes, limit: int, head: int, tail: int, char_in_line: int) -> List[_PrettyToken]:
     """`_tokenize_large_file_content` constructs the intermediate representations. They have no color infomation.
     """
 
     assert head + tail < limit
 
-    def from_line(line: str) -> List[_PrettyToken]:
-        body = line.rstrip()
-        newline = line[len(body):]
-        tokens = []
-        tokens.append(_PrettyToken(_PrettyTokenType.BODY, body))
-        if newline:
-            if newline in ('\n', '\r\n'):
-                tokens.append(_PrettyToken(_PrettyTokenType.NEWLINE, newline))
-            else:
-                whitespace = newline.rstrip('\n')
-                newline = newline[len(whitespace):]
-                if whitespace:
-                    tokens.append(_PrettyToken(_PrettyTokenType.WHITESPACE, whitespace))
-                tokens.append(_PrettyToken(_PrettyTokenType.HINT, '(trailing whitespace)'))
-                if newline:
-                    tokens.append(_PrettyToken(_PrettyTokenType.NEWLINE, newline))
-        return tokens
-
     def candidate_do_nothing(text: str) -> List[_PrettyToken]:
         tokens = []
         for line in text.splitlines(keepends=True):
-            tokens += from_line(line)
+            tokens += _tokenize_line(line)
         return tokens
 
     def candidate_line_based(text: str) -> List[_PrettyToken]:
@@ -63,10 +100,10 @@ def _tokenize_large_file_content(*, content: bytes, limit: int, head: int, tail:
 
         tokens = []
         for line in lines[:head]:
-            tokens += from_line(line)
+            tokens += _tokenize_line(line)
         tokens.append(_PrettyToken(_PrettyTokenType.HINT, '... ({} lines) ...\n'.format(len(lines[head:-tail]))))
         for line in lines[-tail:]:
-            tokens += from_line(line)
+            tokens += _tokenize_line(line)
         return tokens
 
     def candidate_char_based(text: str) -> List[_PrettyToken]:
@@ -77,10 +114,10 @@ def _tokenize_large_file_content(*, content: bytes, limit: int, head: int, tail:
         r = len(text) - char_in_line * tail
         tokens = []
         for line in text[:l].splitlines(keepends=True):
-            tokens += from_line(line)
+            tokens += _tokenize_line(line)
         tokens.append(_PrettyToken(_PrettyTokenType.HINT, '... ({} chars) ...'.format(r - l)))
         for line in text[r:].splitlines(keepends=True):
-            tokens += from_line(line)
+            tokens += _tokenize_line(line)
         return tokens
 
     def count_size(tokens: Iterable[_PrettyToken]) -> int:
@@ -89,35 +126,27 @@ def _tokenize_large_file_content(*, content: bytes, limit: int, head: int, tail:
             size += len(s)
         return size
 
-    if not content:
-        return [_PrettyToken(_PrettyTokenType.HINT, '(empty)')]
-
-    tokens = []
-    try:
-        text = content.decode()
-    except UnicodeDecodeError as e:
-        tokens.append(_PrettyToken(_PrettyTokenType.HINT, str(e)))
-        text = content.decode(errors='replace')
-
-    candidates: List[List[_PrettyToken]] = [
-        candidate_do_nothing(text),
-        candidate_line_based(text),
-        candidate_char_based(text),
-    ]
-    tokens.extend(min(candidates, key=count_size))
-
-    assert len(tokens) >= 1
-    if tokens[-1][0] == _PrettyTokenType.BODY:
-        tokens.append(_PrettyToken(_PrettyTokenType.HINT, '(no trailing newline)'))
-    if not text.rstrip('\n'):
-        tokens.append(_PrettyToken(_PrettyTokenType.HINT, '(only newline)'))
-
+    # Choose the shortest one from the three candidates.
+    tokens, text = _decode_with_recovery(content)
+    if text:
+        candidates: List[List[_PrettyToken]] = [
+            candidate_do_nothing(text),
+            candidate_line_based(text),
+            candidate_char_based(text),
+        ]
+        tokens.extend(min(candidates, key=count_size))
+    tokens = _warn_if_empty(tokens)
     return tokens
 
 
-def _render_tokens_for_large_file_content(*, tokens: List[_PrettyToken], font_bold: Callable[[str], str], font_dim: Callable[[str], str]) -> str:
+def _render_tokens(*, tokens: List[_PrettyToken], font_bold: Optional[Callable[[str], str]] = None, font_dim: Optional[Callable[[str], str]] = None) -> str:
     """`_tokenize_large_file_content` generate the result string. It is colored.
     """
+
+    if font_bold is None:
+        font_bold = lambda s: colorama.Style.BRIGHT + s + colorama.Style.RESET_ALL
+    if font_dim is None:
+        font_dim = lambda s: colorama.Style.DIM + s + colorama.Style.RESET_ALL
 
     result = []
     for key, value in tokens:
@@ -135,17 +164,24 @@ def _render_tokens_for_large_file_content(*, tokens: List[_PrettyToken], font_bo
     return ''.join(result)
 
 
-def make_pretty_large_file_content(content: bytes, limit: int, head: int, tail: int, bold: bool = False) -> str:
+def make_pretty_large_file_content(content: bytes, limit: int, head: int, tail: int) -> str:
     char_in_line, _ = shutil.get_terminal_size()
     char_in_line = max(char_in_line, 40)  # shutil.get_terminal_size() may return too small values (e.g. (0, 0) on Circle CI) successfully (i.e. fallback is not used). see https://github.com/kmyk/online-judge-tools/pull/611
     tokens = _tokenize_large_file_content(content=content, limit=limit, head=head, tail=tail, char_in_line=char_in_line)
+    return _render_tokens(tokens=tokens)
 
-    if bold:
-        font_bold = lambda s: colorama.Style.BRIGHT + s + colorama.Style.RESET_ALL
-    else:
-        font_bold = lambda s: s
-    font_dim = lambda s: colorama.Style.DIM + s + colorama.Style.RESET_ALL
-    return _render_tokens_for_large_file_content(tokens=tokens, font_bold=font_bold, font_dim=font_dim)
+
+def _tokenize_file_content_without_snipping(content: bytes) -> List[_PrettyToken]:
+    tokens, text = _decode_with_recovery(content)
+    for line in text.splitlines(keepends=True):
+        tokens += _tokenize_line(line)
+    tokens = _warn_if_empty(tokens)
+    return tokens
+
+
+def make_pretty_all(content: bytes) -> str:
+    tokens = _tokenize_file_content_without_snipping(content=content)
+    return _render_tokens(tokens=tokens)
 
 
 def _space_padding(s: str, max_length: int) -> str:

--- a/onlinejudge_command/subcommand/download.py
+++ b/onlinejudge_command/subcommand/download.py
@@ -168,7 +168,7 @@ def run(args: argparse.Namespace) -> bool:
         for ext, path, data in iterate_files_to_write(sample, i=i):
             content = ''
             if not args.silent:
-                content = '\n' + pretty_printers.make_pretty_large_file_content(data, limit=40, head=20, tail=10, bold=True)
+                content = '\n' + pretty_printers.make_pretty_large_file_content(data, limit=40, head=20, tail=10)
             logger.info('%sput: %s%s', ext, sample.name, content)
             if not args.dry_run:
                 path.parent.mkdir(parents=True, exist_ok=True)

--- a/onlinejudge_command/subcommand/generate_input.py
+++ b/onlinejudge_command/subcommand/generate_input.py
@@ -79,7 +79,7 @@ def write_result(input_data: bytes, output_data: Optional[bytes], *, input_path:
 
         if print_data:
             logger.info(utils.NO_HEADER + 'input:')
-            logger.info(utils.NO_HEADER + '%s', pretty_printers.make_pretty_large_file_content(input_data, limit=40, head=20, tail=10, bold=True))
+            logger.info(utils.NO_HEADER + '%s', pretty_printers.make_pretty_large_file_content(input_data, limit=40, head=20, tail=10))
         with input_path.open('wb') as fh:
             fh.write(input_data)
         logger.info(utils.SUCCESS + 'saved to: %s', input_path)
@@ -87,7 +87,7 @@ def write_result(input_data: bytes, output_data: Optional[bytes], *, input_path:
         if output_data is not None:
             if print_data:
                 logger.info(utils.NO_HEADER + 'output:')
-                logger.info(pretty_printers.make_pretty_large_file_content(output_data, limit=40, head=20, tail=10, bold=True))
+                logger.info(pretty_printers.make_pretty_large_file_content(output_data, limit=40, head=20, tail=10))
             with output_path.open('wb') as fh:
                 fh.write(output_data)
             logger.info(utils.SUCCESS + 'saved to: %s', output_path)
@@ -194,7 +194,7 @@ def try_hack_once(generator: str, command: str, hack: str, *, tle: Optional[floa
         if conflicted_name is not None:
             submit(logger.warning, 'The same input is already generated at %s. Please use a random input generator.', conflicted_name)
             submit(logger.info, utils.NO_HEADER + 'input:')
-            submit(logger.info, utils.NO_HEADER + '%s', pretty_printers.make_pretty_large_file_content(input_data, limit=40, head=20, tail=10, bold=True))
+            submit(logger.info, utils.NO_HEADER + '%s', pretty_printers.make_pretty_large_file_content(input_data, limit=40, head=20, tail=10))
 
         # generate output
         submit(logger.info, 'generate output...')
@@ -220,9 +220,9 @@ def try_hack_once(generator: str, command: str, hack: str, *, tle: Optional[floa
         expected = output_data.decode()
         if not simple_match(answer, expected):
             logger.info(utils.FAILURE + '' + utils.red('WA'))
-            logger.info(utils.NO_HEADER + 'input:\n%s', pretty_printers.make_pretty_large_file_content(input_data, limit=40, head=20, tail=10, bold=True))
-            logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_large_file_content(answer.encode(), limit=40, head=20, tail=10, bold=True))
-            logger.info(utils.NO_HEADER + 'expected:\n%s', pretty_printers.make_pretty_large_file_content(output_data, limit=40, head=20, tail=10, bold=True))
+            logger.info(utils.NO_HEADER + 'input:\n%s', pretty_printers.make_pretty_large_file_content(input_data, limit=40, head=20, tail=10))
+            logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_large_file_content(answer.encode(), limit=40, head=20, tail=10))
+            logger.info(utils.NO_HEADER + 'expected:\n%s', pretty_printers.make_pretty_large_file_content(output_data, limit=40, head=20, tail=10))
             status = 'WA'
 
         if status == 'AC':

--- a/onlinejudge_command/subcommand/generate_output.py
+++ b/onlinejudge_command/subcommand/generate_output.py
@@ -67,7 +67,7 @@ def generate_output_single_case(test_name: str, test_input_path: pathlib.Path, *
             logger.info('skipped.')
             return
         assert answer is not None
-        logger.info(utils.NO_HEADER + '' + pretty_printers.make_pretty_large_file_content(answer, limit=40, head=20, tail=10, bold=True))
+        logger.info(utils.NO_HEADER + '' + pretty_printers.make_pretty_large_file_content(answer, limit=40, head=20, tail=10))
 
         # find the destination path
         match_result: Optional[Match[Any]] = fmtutils.match_with_format(args.directory, args.format, test_input_path)

--- a/onlinejudge_command/subcommand/submit.py
+++ b/onlinejudge_command/subcommand/submit.py
@@ -80,7 +80,7 @@ def run(args: argparse.Namespace) -> bool:
 
     # report code
     logger.info('code (%d byte):', len(code))
-    logger.info(utils.NO_HEADER + '%s', pretty_printers.make_pretty_large_file_content(code, limit=30, head=10, tail=10, bold=True))
+    logger.info(utils.NO_HEADER + '%s', pretty_printers.make_pretty_large_file_content(code, limit=30, head=10, tail=10))
 
     with utils.new_session_with_our_user_agent(path=args.cookie) as sess:
         # check the login status

--- a/onlinejudge_command/subcommand/test.py
+++ b/onlinejudge_command/subcommand/test.py
@@ -166,7 +166,7 @@ def display_result(proc: subprocess.Popen, answer: str, memory: Optional[float],
         if does_print_input and not is_input_printed:
             is_input_printed = True
             with test_input_path.open('rb') as inf:
-                logger.info(utils.NO_HEADER + 'input:\n%s', pretty_printers.make_pretty_large_file_content(inf.read(), limit=40, head=20, tail=10, bold=True))
+                logger.info(utils.NO_HEADER + 'input:\n%s', pretty_printers.make_pretty_large_file_content(inf.read(), limit=40, head=20, tail=10))
 
     # check TLE, RE or not
     status = JudgeStatus.AC
@@ -199,19 +199,22 @@ def display_result(proc: subprocess.Popen, answer: str, memory: Optional[float],
             else:
                 expected = ''
             if display_mode == DisplayMode.SUMMARY:
-                logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_large_file_content(answer.encode(), limit=40, head=20, tail=10, bold=True))
-                logger.info(utils.NO_HEADER + 'expected:\n%s', pretty_printers.make_pretty_large_file_content(expected.encode(), limit=40, head=20, tail=10, bold=True))
+                logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_large_file_content(answer.encode(), limit=40, head=20, tail=10))
+                logger.info(utils.NO_HEADER + 'expected:\n%s', pretty_printers.make_pretty_large_file_content(expected.encode(), limit=40, head=20, tail=10))
             elif display_mode == DisplayMode.DIFF:
                 if max(answer.count('\n'), expected.count('\n')) <= 40:
                     pretty_printers.display_side_by_side_color(answer, expected)
                 else:
                     pretty_printers.display_snipped_side_by_side_color(answer, expected)
+            elif display_mode == DisplayMode.ALL:
+                logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_all(answer.encode()))
+                logger.info(utils.NO_HEADER + 'expected:\n%s', pretty_printers.make_pretty_all(expected.encode()))
             else:
                 assert False
     if match_result is None:
         if not silent:
             print_input()
-            logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_large_file_content(answer.encode(), limit=40, head=20, tail=10, bold=True))
+            logger.info(utils.NO_HEADER + 'output:\n%s', pretty_printers.make_pretty_large_file_content(answer.encode(), limit=40, head=20, tail=10))
     if status == JudgeStatus.AC:
         logger.info(utils.SUCCESS + '' + utils.green('AC'))
 

--- a/tests/command_test.py
+++ b/tests/command_test.py
@@ -549,6 +549,48 @@ class TestTest(unittest.TestCase):
             }],
         )
 
+    def test_call_test_with_display_mode_all(self):
+        self.snippet_call_test(
+            args=['-c', cat(), '--display-mode', 'all'],
+            files=[
+                {
+                    'path': 'test/sample-1.in',
+                    'data': 'foo\nfoobar\n'
+                },
+                {
+                    'path': 'test/sample-1.out',
+                    'data': 'foo\nfoobar\n'
+                },
+                {
+                    'path': 'test/sample-2.in',
+                    'data': 'bar\nfoobar\n'
+                },
+                {
+                    'path': 'test/sample-2.out',
+                    'data': 'bar\nbarbar\n'
+                },
+            ],
+            expected=[{
+                'status': 'AC',
+                'testcase': {
+                    'name': 'sample-1',
+                    'input': '%s/test/sample-1.in',
+                    'output': '%s/test/sample-1.out',
+                },
+                'output': 'foo\nfoobar\n',
+                'exitcode': 0,
+            }, {
+                'status': 'WA',
+                'testcase': {
+                    'name': 'sample-2',
+                    'input': '%s/test/sample-2.in',
+                    'output': '%s/test/sample-2.out',
+                },
+                'output': 'bar\nfoobar\n',
+                'exitcode': 0,
+            }],
+        )
+
     def test_call_test_select(self):
         self.snippet_call_test(
             args=['-c', cat(), 'test/sample-2.in', 'test/sample-3.in', 'test/sample-3.out'],

--- a/tests/pretty_printers.py
+++ b/tests/pretty_printers.py
@@ -2,8 +2,80 @@
 """
 
 import unittest
+from typing import *
 
-from onlinejudge_command.pretty_printers import _PrettyToken, _PrettyTokenType, _render_tokens_for_large_file_content, _tokenize_large_file_content
+from onlinejudge_command.pretty_printers import _PrettyToken, _PrettyTokenType, _render_tokens, _tokenize_file_content_without_snipping, _tokenize_large_file_content, _tokenize_line
+
+
+class TokenizeLineTest(unittest.TestCase):
+    def test_simple(self) -> None:
+        line = 'hello\n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.BODY, 'hello'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+        ]
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
+
+    def test_crlf(self) -> None:
+        line = 'hello\r\n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.BODY, 'hello'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\r\n'),
+        ]
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
+
+    def test_with_whitespace(self) -> None:
+        line = 'hello  \t\tworld\n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.BODY, 'hello'),
+            _PrettyToken(_PrettyTokenType.WHITESPACE, '  \t\t'),
+            _PrettyToken(_PrettyTokenType.BODY, 'world'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+        ]
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
+
+    def test_without_newline(self) -> None:
+        line = 'hello'
+        expected = [
+            _PrettyToken(_PrettyTokenType.BODY, 'hello'),
+        ]
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
+
+    def test_trailing_whitespace(self) -> None:
+        line = 'hello    \n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.BODY, 'hello'),
+            _PrettyToken(_PrettyTokenType.WHITESPACE, '    '),
+            _PrettyToken(_PrettyTokenType.HINT, '(trailing whitespace)'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+        ]
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
+
+    def test_only_newline(self) -> None:
+        line = '\n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+        ]
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
+
+    def test_empty_string(self) -> None:
+        line = ''
+        expected: List[_PrettyToken] = []
+
+        actual = _tokenize_line(line=line)
+        self.assertEqual(actual, expected)
 
 
 class TokenizeLargeFileContentTest(unittest.TestCase):
@@ -65,8 +137,70 @@ class TokenizeLargeFileContentTest(unittest.TestCase):
         actual = _tokenize_large_file_content(content=content, limit=limit, head=head, tail=tail, char_in_line=char_in_line)
         self.assertEqual(actual, expected)
 
+    def test_empty(self) -> None:
+        content = b''
+        limit = 40
+        head = 20
+        tail = 10
+        char_in_line = 40
+        expected = [
+            _PrettyToken(_PrettyTokenType.HINT, '(empty)'),
+        ]
 
-class RenderTokensForLargeFileContentTest(unittest.TestCase):
+        actual = _tokenize_large_file_content(content=content, limit=limit, head=head, tail=tail, char_in_line=char_in_line)
+        self.assertEqual(actual, expected)
+
+    def test_only_newlines(self) -> None:
+        content = b'\r\n\n'
+        limit = 40
+        head = 20
+        tail = 10
+        char_in_line = 40
+        expected = [
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\r\n'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+            _PrettyToken(_PrettyTokenType.HINT, '(only newline)'),
+        ]
+
+        actual = _tokenize_large_file_content(content=content, limit=limit, head=head, tail=tail, char_in_line=char_in_line)
+        self.assertEqual(actual, expected)
+
+
+class TokenizeFileContentWithoutSnippingTest(unittest.TestCase):
+    def test_small(self) -> None:
+        content = b'hello\nworld\n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.BODY, 'hello'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+            _PrettyToken(_PrettyTokenType.BODY, 'world'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+        ]
+
+        actual = _tokenize_file_content_without_snipping(content=content)
+        self.assertEqual(actual, expected)
+
+    def test_empty(self) -> None:
+        content = b''
+        expected = [
+            _PrettyToken(_PrettyTokenType.HINT, '(empty)'),
+        ]
+
+        actual = _tokenize_file_content_without_snipping(content=content)
+        self.assertEqual(actual, expected)
+
+    def test_only_newlines(self) -> None:
+        content = b'\r\n\n'
+        expected = [
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\r\n'),
+            _PrettyToken(_PrettyTokenType.NEWLINE, '\n'),
+            _PrettyToken(_PrettyTokenType.HINT, '(only newline)'),
+        ]
+
+        actual = _tokenize_file_content_without_snipping(content=content)
+        self.assertEqual(actual, expected)
+
+
+class RenderTokensTest(unittest.TestCase):
     def test_simple(self) -> None:
         tokens = [
             _PrettyToken(_PrettyTokenType.BODY, 'hello'),
@@ -83,22 +217,24 @@ class RenderTokensForLargeFileContentTest(unittest.TestCase):
 
         font_dim = lambda s: '<dim>' + s + '</dim>'
         font_bold = lambda s: '<bold>' + s + '</bold>'
-        actual = _render_tokens_for_large_file_content(tokens=tokens, font_bold=font_bold, font_dim=font_dim)
+        actual = _render_tokens(tokens=tokens, font_bold=font_bold, font_dim=font_dim)
         self.assertEqual(actual, expected)
 
     def test_complicated(self) -> None:
         tokens = [
             _PrettyToken(_PrettyTokenType.BODY, 'hello world'),
             _PrettyToken(_PrettyTokenType.WHITESPACE, ' \t'),
+            _PrettyToken(_PrettyTokenType.HINT, 'this is a hint message'),
             _PrettyToken(_PrettyTokenType.NEWLINE, '\r\n'),
         ]
         expected = ''.join([
             '<bold>hello world</bold>',
             '<dim>_\\t</dim>',
+            '<dim>this is a hint message</dim>',
             '<dim>\\r\n</dim>',
         ])
 
         font_dim = lambda s: '<dim>' + s + '</dim>'
         font_bold = lambda s: '<bold>' + s + '</bold>'
-        actual = _render_tokens_for_large_file_content(tokens=tokens, font_bold=font_bold, font_dim=font_dim)
+        actual = _render_tokens(tokens=tokens, font_bold=font_bold, font_dim=font_dim)
         self.assertEqual(actual, expected)

